optin/reporting/test/controllers/completedController.spec.js

```diff
@@ -0,0 +1,191 @@
+			pageSize: 10, pageNumber: 1, optinType: 0, nemAddress: '', symbolAddress: '', transactionHash: ''
+		});
+		res = TestUtils.mockResponse();
+	});
+
+	afterEach(() => {
+		req = {};
+		res = {};
+		restore();
+	});
+
+	describe('getCompleted', () => {
+		const runBasicCompletedDataTests = async (mockNemSources, mockSymbolSources) => {
+			// Arrange:
+			const recordSize = 10;
+			const mockDb = TestUtils.mockCompletedDBRecord(recordSize, mockNemSources, mockSymbolSources);
+
+			getCompletedPaginationStub.returns(Promise.resolve(mockDb));
+			getTotalRecordStub.returns(Promise.resolve(recordSize));
+
+			// Act:
+			await CompletedController.getCompleted(req, res);
+
+			// Assert:
+			const { pagination, data } = res.json.getCall(0).firstArg;
+			const result = data[0];
+
+			expect(pagination).to.be.eql({ pageSize: recordSize, pageNumber: 1, totalRecord: recordSize });
+			expect(data.length).to.be.equal(10);
+
+			expect(result.optin_id).to.be.equal(1);
+			expect(result.isPostoptin).to.be.equal(1);
+			expect(result.label).to.be.eql(mockNemSources.map(nemInfo => nemInfo.label));
+			expect(result.nemAddress).to.be.eql(mockNemSources.map(nemInfo =>
+				new NemFacade.Address(hexStringToByte(nemInfo.address)).toString()));
+			expect(result.nemHeights).to.be.eql(mockNemSources.map(nemInfo => formatStringSplit(nemInfo.height)));
+			expect(result.nemHashes).to.be.eql(mockNemSources.map(nemInfo => formatStringSplit(nemInfo.hashes)));
+			expect(result.nemTimestamps).to.be.eql(mockNemSources.map(nemInfo => formatStringSplit(nemInfo.timestamps)));
+			expect(result.nemBalance).to.be.eql(mockNemSources.map(nemInfo => nemInfo.balance));
+			expect(result.symbolAddress).to.be.eql(mockSymbolSources.map(symbolInfo =>
+				new SymbolFacade.Address(hexStringToByte(symbolInfo.address)).toString()));
+			expect(result.symbolHeights).to.be.eql(mockSymbolSources.map(symbolInfo => symbolInfo.height));
+			expect(result.symbolHashes).to.be.eql(mockSymbolSources.map(symbolInfo => symbolInfo.hashes));
+			expect(result.symbolTimestamps).to.be.eql(mockSymbolSources.map(symbolInfo => symbolInfo.timestamps));
+			expect(result.symbolBalance).to.be.eql(mockSymbolSources.map(symbolInfo => symbolInfo.balance));
+		};
+
+		it('should return basic data', async () => {
+			const mockNemSources = [{
+				address: '682FAFBA20454869F0278748FD9790CFFCE35E8722647B4039',
+				balance: 79268194338335,
+				hashes: 'CDB0AF349823F1638E032DCAB15FD33049C383FEEE642CBC67E11020DC30F190',
+				height: 3004351,
+				label: 'Bithumb',
+				timestamps: '1609403596'
+			}];
+
+			const mockSymbolSources = [{
+				address: '688ADC5D31F49F918AC71DCE18E7085944AEAB0AF60F8AD7',
+				balance: 79268194338335,
+				hashes: 'E49B240D76DAE7277089C2BDA66B297A05AE700361ED253ED48435E9AF9B0FE1',
+				height: 1,
+				timestamps: 1615853185
+			}];
+
+			await runBasicCompletedDataTests(mockNemSources, mockSymbolSources);
+		});
+
+		it('should return data with optin type merge', async () => {
```

Comment:
why is this merge? isn't it 1:1 like previous?

Reply:
nope, this one is consist of multiple hashes in the transaction.

`1:1`
![image](https://user-images.githubusercontent.com/5649156/168020248-8c1b6d91-edaf-4949-a42f-cd230f88bc03.png)

`merge`
![image](https://user-images.githubusercontent.com/5649156/168020342-70aec645-6cc5-4fc3-889b-69c804c06397.png)